File: Charts/piechart.cpp
/*
 This file is part of Marb.
    Marb is free software: you can redistribute it and/or modify
    it under the terms of the Lesser GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License.
    Marb is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    Lesser GNU General Public License for more details.
    You should have received a copy of the Lesser GNU General Public License
    along with Marb.    If not, see <http://www.gnu.org/licenses/>.
 Marb    Copyright (C) 2013    Dimitry Ernot & Romha Korev
*/


#include "piechart.h"
#include "../kernel/global.h"
#include "abstractchart.h"

#include <QDebug>

#include <QPainter>
#include <QPaintEvent>

namespace Marb {

PieChart::PieChart( QWidget* parent ) :
    AbstractChart( new PieChartPrivate(), parent ) {
    this->setEditTriggers( QAbstractItemView::NoEditTriggers );
}


qreal PieChart::startAngle() const {
    const Q_D( PieChart );
    return d->startAngle;
}


void PieChart::configureColor(QPainter &painter, QColor base, int flag ) const {
    switch ( flag ) {
        case 0:
        default:
            painter.setPen( QPen( base.darker( 105 ), 4 ) );
            painter.setBrush( base );
        break;
        case 1:
            painter.setPen( QPen( base.darker( 105 ), 2 ) );
            painter.setBrush( base );
        break;
        case 2:
            base.setAlpha( base.alpha() * 0.75 );
            painter.setPen( QPen( base.lighter( 110 ), 2 ) );
            painter.setBrush( base.lighter( 120 ) );
    }
}


QPainterPath PieChart::itemPart( qreal angle, qreal delta, bool splitted ) const {
    const Q_D( PieChart );
    QPainterPath part;
    part.moveTo( d->rect.center() );
    part.arcTo( d->rect, -angle, -delta );
    if ( splitted == true ) {
        QPointF p = this->splittedOffset( angle, delta );
        part.translate( p.x(), p.y() );
    }
    part.closeSubpath();
    if ( d->ring == true ) {
        QPainterPath p;
        p.addEllipse( d->rect.center(), d->rect.width() * 0.3, d->rect.height() * 0.3 );
        part = part.subtracted( p );
    }
    return part;
}


QPainterPath PieChart::itemPath( const QModelIndex& index ) const {
    const Q_D( PieChart );
    QPainterPath path;
    qreal angle = d->startAngle;
    for ( int r = 0; r < index.row(); ++r ) {
        QModelIndex id = this->model()->index( r, 0 );
        qreal v = qAbs( this->model()->data( id ).toReal() );
        qreal delta = 360.0 * v/d->total;
        angle += delta;
    }
    qreal v = qAbs( this->model()->data( index ).toReal() );
    qreal delta = 360.0 * v/d->total;
    if ( this->selectionModel()->selectedIndexes().contains( index )) {
        path = this->itemPart( angle, delta, true );
    } else {
        path = this->itemPart( angle, delta );
    }
    return path;
}


void PieChart::paintChart( QPainter &painter ) const {
    const Q_D( PieChart );
    //this->updateChart();
    painter.save();
    int rows = this->model()->rowCount();
    qreal angle = d->startAngle;
    for ( int i = 0; i < rows; ++i ) {
        QModelIndex index = this->model()->index( i, 0 );
        QColor color( this->model()->data( index, Qt::DecorationRole ).toString() );
        if ( !color.isValid() ) {
            color = Color::predefinedColor( i );
        }
        qreal v = qAbs( this->model()->data( index ).toReal() );
        qreal delta = 360.0 * v/d->total;
        bool isSelected = this->selectionModel()->selectedIndexes().contains( index )
                                            || this->currentIndex() == index;
        if ( d->splitted == false ) {
            this->paintPart( painter, angle, delta, color, isSelected );
        } else {
            this->paintPartSplitted( painter, angle, delta, color, isSelected );
        }
        angle += delta;
    }
    painter.drawText( d->titleRect, Qt::AlignHCenter | Qt::TextWordWrap, d->title );

    painter.restore();
}


/*Paint the legend in the QRect self._legendRect
The legend corresponds to the text in the Horizontal QHeaderView and the style defined for each column.
*/
void PieChart::paintLegend( QPainter& painter) const {
    const Q_D( PieChart );
    painter.save();
    QFontMetrics metrics( this->font() );
    int metricsH = metrics.height();
    int h = metricsH + 10;
    int rows = this->model()->rowCount();
    int w = 0;
    int maxWidth = d->legendRect.width();
    QPoint legendPos( d->legendRect.topLeft() );
    QPoint pos = legendPos + QPoint( 50, 0);
    for (int r = 0; r < rows; ++r ) {
        QString s( this->model()->headerData( r, Qt::Vertical ).toString() );
        int sWidth = metrics.width( s ) + 50;
        QPoint p;
        if ( ( w + sWidth ) > maxWidth ) {
            int y = pos.y();
            p = QPoint( d->legendRect.x(), y + h );
            pos = QPoint( d->legendRect.x(), y + h );
            w = sWidth;
            pos += QPoint( sWidth, 0 );
        } else {
            p = pos + QPoint( -40,    0 );
            w += sWidth;
            pos += QPoint( sWidth, 0 );
        }
        this->paintSerieLegend(painter, r, p, metricsH);
    }
    painter.restore();
}


void PieChart::paintSerieLegend( QPainter& painter, int serie, QPoint pos, int metricsH ) const {

    QColor color( this->model()->data( this->model()->index( serie, 0 ), Qt::DecorationRole ).toString() );
    if ( !color.isValid() ) {
        color = Color::predefinedColor( serie );
    }

    QRect r( pos.x(), pos.y() - 20, 14, 14 );
    QPoint posText = pos + QPoint( 20, -metricsH/2 );

    QString s = this->model()->headerData( serie, Qt::Vertical ).toString();
    painter.drawText( posText, s );
    painter.save();
    painter.setBrush( color );
    painter.setPen( QPen( color.darker( 100 ), 1 ) );
    painter.drawRect( r );

    painter.restore();
}

void PieChart::paintPart( QPainter& painter, qreal angle, qreal delta, QColor color, bool isSelected ) const {
    if ( isSelected == true ) {
        this->paintPartSplitted( painter, angle, delta, color );
        return;
    }
    QPainterPath part = this->itemPart( angle, delta );
    painter.save();
    painter.setClipPath( part ); /* To avoid the "borders superposition" */
    int flag = 0;
    this->configureColor( painter, color, flag );
    painter.drawPath( part );
    painter.restore();
}


void PieChart::paintPartSplitted( QPainter &painter, qreal angle, qreal delta,
                                  QColor color, bool isSelected ) const {
    const Q_D( PieChart );
    QPainterPath part = this->itemPart( angle, delta, true );
    painter.save();
    if ( d->splitted == true
             && ( !this->selectionModel()->selectedIndexes().isEmpty() || this->currentIndex().isValid() )
             && isSelected == false ) {
        this->configureColor( painter, color, 2 );
    } else {
        this->configureColor( painter, color, 1 );
    }
    painter.drawPath( part );
    painter.restore();
}


void PieChart::scrollTo(const QModelIndex &/*index*/, ScrollHint /*hint*/) {
}


void PieChart::setRing( bool ring ) {
    Q_D( PieChart );
    d->ring = ring;
}


void PieChart::setSplitted( bool splitted ) {
    Q_D( PieChart );
    d->splitted = splitted;
}


void PieChart::setStartAngle( qreal angle ) {
    Q_D( PieChart );
    d->startAngle = angle;
    this->viewport()->update();
}


QPointF PieChart::splittedOffset( qreal angle, qreal delta ) const {
    const Q_D( PieChart );
    QPainterPath part;
    part.moveTo( d->rect.center() );
    part.arcTo( d->rect, -angle, -delta );
    part.closeSubpath();
    QPointF p = part.pointAtPercent( 0.5 );
    QLineF line( p, d->rect.center() );
    line.setLength( line.length() * 1.1 );
    p = line.p2();
    return d->rect.center() - p;
}


void PieChart::updateRects() {
    if ( this->model() == 0 )    {
        return;
    }
    this->defineRects();
    //myTitleRect.moveTo( axis()->myValuesRect.bottomLeft() );
    //myTitleRect.translate( (axis()->myValuesRect.width() - myTitleRect.width())/2, 20 );
}


QRect PieChart::visualRect( const QModelIndex& index ) const {
    Q_UNUSED(index)
    /* To force to repaint the whole chart. Not only the area of the part*/
    return QRect( 0, 0, width(), height() );
}


int PieChart::scan() {
    /*Scans values in the model to find the minimum and the maximum. Returns the width needed to display the Y scale.
    If the values are greater than zero, the minimum is equal to 0. If the values are less than 0, the maximum is equal to 0.
    If a value is not a number (undefined, a string, etc.), she's considered as equal to 0. */

    Q_D( PieChart );

    int rows = this->model()->rowCount();

    d->total = 0;
    for ( int r = 0; r < rows; ++r ) {
        QString s( this->model()->headerData( r, Qt::Vertical ).toString() );

        qreal value = this->model()->index( r, 0 ).data().toDouble();
        d->total += value;
    }
    return 0;
}


void PieChart::defineRects() {
    Q_D( PieChart );
    d->rect = this->contentsRect();
    this->calculateLegendRect( d->rect );
    d->rect.setWidth( d->rect.width() - d->legendRect.width() - 10 );
    
    if ( d->title != "" ) {
        QFont font = this->font();
        font.setItalic( true );
        QFontMetrics metrics( font );
        QRect r( 0, 0, d->rect.width() - 40, 0 );
        d->titleRect = metrics.boundingRect( r, Qt::AlignHCenter | Qt::AlignTop | Qt::TextWordWrap, d->title );
        //myAxis->chartRect.setHeight( myAxis->chartRect.height() - myTitleRect.height() - 20 );
        d->rect.setHeight( d->rect.height() - d->titleRect.height() - 20 );
    }


    int w = qMin( d->rect.width() * 0.80, d->rect.height() * 0.80 );

    d->rect.setWidth( w );
    d->rect.setHeight( w );

    d->titleRect.moveTo( d->titleRect.x(), d->rect.top() );
    d->rect.moveTo( d->rect.x(), d->titleRect.bottom() + 10 );
    d->legendRect.moveTo( d->rect.right() + 10, d->rect.top() );
    
}


void PieChart::calculateLegendRect( const QRect& source ) {
    Q_D( PieChart );
    QFontMetrics metrics( font() );
    int h = metrics.height() + 10;
    int rows = this->model()->rowCount();

    int w = 0;

    for( int r = 0; r < rows; ++r ) {
        QString s( this->model()->headerData( r, Qt::Vertical ).toString() );
        w = qMax( metrics.width( s ) + 40, w );
    }

    w = qMin( source.width(), w );
        
    d->legendRect = QRect( source.right() - w, this->contentsMargins().top(), w, rows * h );
}

}

File: Marb/Charts/PieChart.py
from Marb.Global import Color, Render
from PySide.QtGui import QAbstractItemView, QPen, QRegion, QColor, QFontMetrics, QPainter, QPixmap, QPainterPath
from PySide.QtCore import QRect, QPointF, Qt, QModelIndex, QLineF
#from ..MarbAbstractItemView import MarbAbstractItemView
from .Chart import Chart
class PieChart( Chart ):
	class ColorConf:
		Normal = 0
		Lighter = 1
		Darker = 2
		Translucent = 4


	def __init__(self, parent=None):
		''' Constructor. Constructs an instance with the given [parent].
		parent: QWidget
		'''
		super(PieChart, self).__init__( parent )
		self.setEditTriggers( QAbstractItemView.NoEditTriggers )
		self._Ring = False
		self._Splitted = False
		self._StartAngle = 0.0
		self._Legend = ""
		self._Render = Render.Plain
		self._chartRect = QRect()
		self._valuesRect = QRect()
		self._ellipseFactor = 0.80


	def configureColor(self, painter, base, flag):
		''' Sets the pen and brush colors for the painter according to the bas color and the PieChart.ColorConf flags.
		'''
		painter.setPen( QPen( base, 3 ) )
		c = base
		if flag == PieChart.ColorConf.Darker:
			c = base.darker( 110 )
		elif flag == PieChart.ColorConf.Lighter:
			c = base.lighter( 120 )
		if self._Render == Render.Plain:
			painter.setPen( QPen( base.darker( 110 ), 2 ) )
		if self._Render == Render.Translucent:
				c.setAlpha( 180 )
		if self._Render == Render.Wireframe:
				c.setAlpha( 10 )
		painter.setBrush( c )


	def itemPart( self, angle, delta, splitted = False ):
		''' Returns a QPainterPath representing the pie part for the given angle and delta.
		'''
		part = QPainterPath()
		part.moveTo( self._valuesRect.center() )
		part.arcTo( self._valuesRect, -angle, -delta )
		if splitted == True:
			(x, y) = self.splittedOffset(angle, delta, True)
			part.translate( x, y )
		part.closeSubpath()
		if self._Ring == True:
			p = QPainterPath()
			p.addEllipse( self._valuesRect.center(), self._valuesRect.width() * 0.2, self._valuesRect.height() * 0.2 )
			part = part.subtracted( p )
		return part


	def itemPath( self, index ):
		''' Returns a QPainterPath representing the pie part for the given index.
		'''
		path = QPainterPath()
		if index.column() != 0:
			return path
		angle = self._StartAngle
		for r in range( 0, index.row() ):
			idx = self.model().index( r, 0 )
			v = abs( self.model().data( idx ) )
			delta = 360.0 * v/self._Total
			angle += delta
		v = abs( self.model().data( index ) )
		delta = 360.0 * v/self._Total
		if index in self.selectionModel().selectedIndexes():
			path = self.itemPart( angle, delta, True )
		else :
			path = self.itemPart( angle, delta )
		return path


	def paintChart( self, painter ):
		rows = self.model().rowCount()
		self.updateValues()
		painter.save()
		painter.setRenderHint( QPainter.Antialiasing )
		angle = self._StartAngle
		self._leftLabels = []
		self._rightLabels = []
		selectedIndexes = self.selectionModel().selectedIndexes()
		for i in range( 0, rows ):
			index = self.model().index( i, 0 )
			# Detect if model contains a specific color. Otherwise, use a predefined color.
			color = QColor( self.model().data( index, Qt.DecorationRole ) )
			if not color.isValid():
				color = Color.colorAt( i )
			value = 0
			try:
				value = float(self.model().data( index ))
			except:
				value = 0
			v = abs( value )
			delta = 360.0 * v/self._Total
			centerAngle = angle + delta/2.0
			if centerAngle < 90: # Right side
				self._rightLabels.append( (i, centerAngle, color) )
			elif centerAngle > 270: # Right side
				self._rightLabels.insert( 0, (i, centerAngle, color) )
			else: # Left side
				self._leftLabels.append( (i, centerAngle, color) )
			# A part can be splitted if self._Splitted is True or if self._Splitted is False and the value is selected in the model.
			isSelected = False
			if selectedIndexes != []:
				for idx in selectedIndexes:
					if idx == index:
						isSelected = True
						break
			if self._Splitted == False:
				if isSelected == False:
					self.paintPart( painter, angle, delta, color )
				else:
					self.paintPartSplitted( painter, angle, delta, color )
			else:
				if selectedIndexes != [] and isSelected == False:
					c = QColor( color )
					c.setAlpha( c.alpha() * 0.5 )
					self.paintPartSplitted( painter, angle, delta, c )
				else:
					self.paintPartSplitted( painter, angle, delta, color )
			angle += delta
		painter.drawText( 10, self._chartRect.bottomLeft().y(),
						self.width() - 20, self.height() - self._valuesRect.height(), Qt.AlignHCenter | Qt.TextWordWrap, self._Legend )
		painter.restore()


	def paintEvent( self, event ) :
		if ( self.model() == 0 ) :
			return None
		painter = QPainter( self.viewport() )
		self.paintChart( painter )


	def paintLabels(self, painter):
		''' Paints label for each part.
		'''
		# Paint labels
		metrics = QFontMetrics( self.font() )
		ell = QPainterPath()
		if self._Splitted == False:
			ell.addEllipse( self._valuesRect )
		else:
			ell.addEllipse( self._chartRect )
		l = []
		j = 1
		self._leftLabels.reverse()
		while j < len(self._leftLabels):
			(i, centerAngle, color) = self._leftLabels[j]
			prevCenterAngle = self._leftLabels[j - 1][1]
			y1 = ell.pointAtPercent( centerAngle/360.0 ).y()
			y2 = ell.pointAtPercent( prevCenterAngle/360.0 ).y()
			d = abs(y2 - y1)
			y = 0
			if d < (metrics.height() + 2):
				y = y1 + metrics.height() - d + 20
			else:
				y = y1
			l.append( (i, centerAngle, y, color) )
			j += 1
		for (i, centerAngle, y, color) in l:
			textPos = QPointF(0,0)
			p1 = QPointF(0,0)
			text = str(self.model().headerData( i, Qt.Vertical ))
			p = ell.pointAtPercent( centerAngle/360.0 )
			textPos = QPointF( self._leftLabelRect.x(), y )
			p1 = QPointF(textPos.x(), p.y() )
			r = QRect( textPos.x(), textPos.y() - metrics.height(), self._leftLabelRect.width(), metrics.height() )
			painter.drawText( r, text )
			pen = painter.pen()
			painter.setPen( color )
			painter.drawLine( p1, p )
			painter.setPen( pen )


	def paintPart( self, painter, angle, delta, color ):
		''' Paints the part defined by angle and delta with the given color.
		'''
		part = self.itemPart( angle, delta )
		painter.save()
		painter.setClipPath( part ) # To avoid the "borders superposition"
		self.configureColor(painter, color, PieChart.ColorConf.Lighter)
		painter.drawPath( part )
		painter.restore()


	def paintPartSplitted( self, painter, angle, delta, color ):
		''' Does the same work than paintPart but with an offset.
		Used to paint parts when the flag splitted is set.
		'''
		part = self.itemPart( angle, delta, True )
		painter.save()
		self.configureColor(painter, color, PieChart.ColorConf.Lighter)
		painter.drawPath( part )
		painter.restore()


	def save( self, filename ):
		pix = QPixmap( self.size() )
		pix.fill( Qt.white )
		painter = QPainter ( pix )
		self.paintChart( painter )
		return pix.save( filename )

	def scrollTo( self, index, hint ):
		pass


	def setLegend( self, legend ) :
		self._Legend = legend


	def setRing( self, ring ) :
		'''Defines the pie chart as a ring (paints a 'hole' at the center).
		'''
		self._Ring = ring


	def setSplitted( self, splitted ) :
		''' sets the flag self._Splitted to splitted.
		If splitted is True, 
		'''
		self._Splitted = splitted


	def splittedOffset( self, angle, delta, splitted ):
 		if splitted == False:
 			return ( 0, 0 )
 		line = QLineF( QPointF(0,0), QPointF((self._valuesRect.width()/2), 0) )
 		line.setAngle( -angle - delta/2 )
 		p1 = line.p2()
 		line.setLength( (line.length() * self._chartRect.width() / self._valuesRect.width()) )
 		p2 = line.p2()
 		return ( (p2.x() - p1.x()), (p2.y() - p1.y()) * 0.80 )


	def _updateRects(self):
		if self.model() == None:
			return None
		self.defineRects()
		self._titleRect.moveTo( self._valuesRect.bottomLeft() )
		self._titleRect.translate( (self._valuesRect.width() - self._titleRect.width())/2, 20 )


	def updateValues( self ):
		metrics = QFontMetrics( self.font() )
		maxLabelWidth = 0
		self._Total = 0
		self._updateRects()
		for i in range( 0, self.model().rowCount() ):
			index = self.model().index( i, 0 )
			value = 0
			try:
				value = float(self.model().data( index ))
			except:
				value = 0
			self._Total += abs( value )
			text = str(self.model().headerData( i, Qt.Vertical ))
			maxLabelWidth = max( metrics.width( text ), maxLabelWidth )
		w = self._chartRect.width()
		h = self._chartRect.height()
		if w < h:
			self._chartRect.setHeight( w )
			self._chartRect.translate( 0, (h- w)/2 )
		else:
			self._chartRect.setWidth( h )
			self._chartRect.translate( (w - h)/2, 0 )
		self._titleRect.moveTo( self._titleRect.x(), self._chartRect.bottom() + 10 )
		w = self._chartRect.width() * 0.80
		h = self._chartRect.height() * 0.80
		dw = self._chartRect.width() - w
		dh = self._chartRect.height() - h
		self._valuesRect = QRect( self._chartRect.x() + dw/2.0, self._chartRect.y() + dh/2.0, w, h )


	def visualRect( self, index ):
		# To force to repaint the whole chart. Not only the area of the part
		return QRect( 0, 0, self.width(), self.height() )


	def visualRegionForSelection( self, selection ):
		region = QRegion()
		region += QRect( 0, 0, self.width(), self.height() )
		return region


